crates/vm/src/system/memory/tree/mod.rs

```diff
@@ -95,19 +118,34 @@ impl<const CHUNK: usize, F: PrimeField32> MemoryNode<CHUNK, F> {
         memory: &MemoryImage<F>,
         hasher: &impl Hasher<CHUNK, F>,
     ) -> MemoryNode<CHUNK, F> {
-        // Construct a BTreeMap that includes the address space in the label calculation,
+        // Construct a Vec that includes the address space in the label calculation,
         // representing the entire memory tree.
-        let mut memory_partition = BTreeMap::new();
+        let mut memory_partition: Vec<(u64, [F; CHUNK])> = Vec::new();
         for ((address_space, pointer), value) in memory.items() {
```

Comment:
Note: It's a bit unfortunate that `MemoryImage` is `(address_space, pointer) -> F` rather than `(address_space, label) -> [F; 8]`. I feel like this means we have to do a lot of extra work here and at the end of a segment to convert it into both forms.